core/model/modx/processors/resource/trash/purge.class.php

```diff
@@ -0,0 +1,210 @@
+
+        //$this->modx->log(MODx::LOG_LEVEL_INFO, "Purging resources: " . implode(',', $this->ids));
+        // fire before empty trash event
+        $this->modx->invokeEvent('OnBeforeEmptyTrash', array(
+            'ids'       => &$this->ids,
+            'resources' => &$this->resources,
+        ));
+
+        //reset($this->resources);
+
+        // we track success and failure independently, as we don't want
+        // to stop in case of single files failing
+        $success = array();
+
+        $this->failures = array(); // we are no more interested in the previous failures, as they are already filtered out
+
+        $permissionsForPurge = array(
+            'save'   => true,
+            'delete' => true,
+        );
+
+        /** @var modResource $resource */
+        foreach ($this->resources as $resource) {
+            if (!$resource->checkPolicy($permissionsForPurge)) {
+                continue;
+            }
+
+            $id = $resource->get('id');
+
+            $resourceGroupResources = $resource->getMany('ResourceGroupResources');
+            $templateVarResources = $resource->getMany('TemplateVarResources');
+
+            /** @var modResourceGroupResource $resourceGroupResource */
+            foreach ($resourceGroupResources as $resourceGroupResource) {
+                $resourceGroupResource->remove();
+            }
+
+            /** @var modTemplateVarResource $templateVarResource */
+            foreach ($templateVarResources as $templateVarResource) {
+                $templateVarResource->remove();
+            }
+
+            // TODO isn't that a problem here?
+            // If resource remove now fails we already removed the tvs!
+            // shouldn't resource->remove also take care of the tvs and resource groups?
+            if ($resource->remove()==false) {
+                // we just add the id to the failures here (we may already have failures from the init with permissions)
+                $this->failures[] = $id;
+            } else {
+                $success[] = $id;
+            }
+        }
+
+        $this->modx->invokeEvent('OnEmptyTrash', array(
+            'num_deleted' => sizeof($success),
+            'resources'   => &$this->resources,
+            'ids'         => &$success,
+        ));
+
+        $this->modx->logManagerAction('empty_trash', 'modResource', implode(',', $success));
+
+        // if nothing was successfully purged, we throw a failure here
+        if (count($this->failures)>0 && count($success)==0) {
+            return $this->failure($this->modx->lexicon('trash.purge_err_delete', array(
+                // TODO get the pagetitles here
+                'list' => implode(',',$this->failures),
+                'count' => count($this->failures)
+            )));
+        }
+
+        if (count($this->failures)==0 && count($success)==0) {
+            return $this->success($this->modx->lexicon('trash.purge_err_nothing'));
+        }
+
+        $msg = "no message";
```

Comment:
Better to just set it to an empty string if there's no message, as semantically, `no message`, is a message.